components/hab/src/command/pkg/uninstall/uninstall_impl.rs

```diff
@@ -61,28 +84,30 @@ pub async fn uninstall_all_but_latest<U>(ui: &mut U,
                                          execution_strategy: ExecutionStrategy,
                                          scope: Scope,
                                          excludes: &[PackageIdent],
+                                         uninstall_hook_mode: UninstallHookMode,
                                          safety: UninstallSafety)
                                          -> Result<usize>
     where U: UIWriter
 {
     let ident = ident.as_ref();
     let mut idents = list::package_list(&ident.clone().into())?;
-    if number_latest_to_keep >= idents.len() {
+    let len = idents.len();
+    if number_latest_to_keep >= len {
         ui.begin(format!("Uninstalling {}", ident))?;
-        ui.status(Status::Skipping,
-                  format!("Only {} packages installed", idents.len()))?;
+        ui.status(Status::Skipping, format!("Only {} packages installed", len))?;
         ui.end(format!("Uninstall of {} complete", ident))?;
         return Ok(0);
     }
-    // Reverse sort the idents so the latest occur first in the list
-    idents.sort_unstable_by(|a, b| b.by_parts_cmp(a));
-    let to_uninstall = &idents[number_latest_to_keep..];
+    // sort the idents so the latest occur last in the list
+    idents.sort_unstable_by(|a, b| a.by_parts_cmp(b));
+    let to_uninstall = &mut idents[..(len - number_latest_to_keep)];
```

Comment:
This is functionally identical, in that the same identifiers to install are targeted, but their ordering is now oldest to most recent... is that significant? It would be nice to add a bit to the comment to say _why_ that's important.

Additionally, if the ordering is important, it would be good for `uninstall_many` to actually enforce that requirement.

[LATER EDIT]: Reading further through the code, my guess is that by ordering them in this way, the last package to be deleted is the most recent, and if that is also the last instance of that package on disk, then that means that the "most recent uninstall hook" is the one that runs. The assumption here is that the latest version of the hook should be able to "undo" anything that previous versions may have done. That makes sense, but it's a pretty subtle point that is really easy to miss in this code. Capturing that requirement both documentation-wise as well as structurally within the code (as well as in a test!) would be very helpful.

Reply:
In terms of the sorting implementation. It seems like it would be best to revert the above bit to what is way and then have `uninstall_many` do a second sort. That does introduce additional cost of a second sort but in this case I don't anticipate that cost being burdensome.